### PR TITLE
Added a method that would get all the files in a directory, validate …

### PR DESCRIPTION
**Pull request by JessicaWNava**
…them, and return a boolean saying if they all pass validation and logs what failed if there is a failure

Test Steps:
1. Make sure the tests pass.

## Changes
- Added a method that would get all the files in a directory, validate them, and return a boolean saying if they pass validation. An error is logged if there is a failure. 

### Testing
- [x] Tested locally?
- [ ] Ran `./prime test` or `./gradlew testSmoke` against local Docker ReportStream container?
- [ ] (For Changes to /frontend-react/...) Ran `npm run lint:write`? 
- [x] Added tests?

### Process
- [ ] Are there licensing issues with any new dependencies introduced?
- [ ] Includes a summary of what a code reviewer should test/verify?
- [ ] Updated the release notes?
- [ ] Database changes are submitted as a separate PR?
- [ ] DevOps team has been notified if PR requires ops support?

## Linked Issues
#issue 10492

### CODE REVIEW

File: prime-router/src/main/kotlin/azure/BlobAccess.kt [hunk 1]
```diff
@@ -234,6 +242,75 @@ class BlobAccess() : Logging {
...
+                        FhirTransformSchema::class.java
+                    )
+                    val fhirTransform = if (configSchema is FhirTransformSchema) {
+                        FhirTransformer(configSchema).transform(inputBundle)
+                    } else {
+                        throw InternalServerErrorException("Transform schema $currentTransformName not a FhirTransform")
+                    }
+                    val output = sourceContainer.getBlobClient(path + "output.fhir").downloadContent()
+                    // Unfortunately need to do this to get it to be the same format.
+                    val finalOutput = FhirTranscoder.encode(FhirTranscoder.decode(output.toString()))
```
Comment (arnejduranovic) [inline prime-router/src/main/kotlin/azure/BlobAccess.kt:287]:
```suggestion
                    val expectedOutput = FhirTranscoder.encode(FhirTranscoder.decode(output.toString()))
```

File: prime-router/src/main/kotlin/azure/BlobAccess.kt [hunk 2]
```diff
@@ -234,6 +242,75 @@ class BlobAccess() : Logging {
...
+                    val finalTransform = FhirTranscoder.encode(fhirTransform)
```
Comment (arnejduranovic) [inline prime-router/src/main/kotlin/azure/BlobAccess.kt:288]:
```suggestion
                    val transformedInputString = FhirTranscoder.encode(fhirTransform)
```

File: prime-router/src/main/kotlin/azure/BlobAccess.kt [hunk 3]
```diff
@@ -234,6 +242,75 @@ class BlobAccess() : Logging {
...
+                val inputBlobClient = sourceContainer.getBlobClient(path + "input.fhir")
+                val input = inputBlobClient.downloadContent()
+                val inputBundle = FhirTranscoder.decode(input.toString())
+
+                if (schemaType == SchemaType.FHIR_FHIR) {
+                    val configSchema = ConfigSchemaReader.readOneYamlSchema(
+                        transform.toStream(),
+                        FhirTransformSchema::class.java
+                    )
+                    val fhirTransform = if (configSchema is FhirTransformSchema) {
```
Comment (arnejduranovic) [inline prime-router/src/main/kotlin/azure/BlobAccess.kt:280]:
```suggestion
                    val transformedInputBundle = if (configSchema is FhirTransformSchema) {
```

File: prime-router/src/main/kotlin/azure/BlobAccess.kt [hunk 4]
```diff
@@ -234,6 +242,75 @@ class BlobAccess() : Logging {
...
+                    // Unfortunately need to do this to get it to be the same format.
+                    val finalOutput = FhirTranscoder.encode(FhirTranscoder.decode(output.toString()))
+                    val finalTransform = FhirTranscoder.encode(fhirTransform)
+                    // .equals is required here == does not compare it properly, despite what IntelliJ says
+                    if (!finalTransform.equals(finalOutput)) {
+                        logger.error("Validation failed for transform $currentTransformName")
+                        return false
+                    }
+                } else {
+                    val converterSchema = ConfigSchemaReader.readOneYamlSchema(transform.toStream())
```
Comment (arnejduranovic) [inline prime-router/src/main/kotlin/azure/BlobAccess.kt:295]: The way `ConfigSchemaReader` is designed to be used is via `fromFile`. `fromFile` will merge schemas that are referenced into a final schema, and since we are calling `readOneYamlSchema`, we end up skipping that functionality... I think we will need to figure out how to resolve this issue, unless the design notes referenced schemas will not be supported?

File: prime-router/src/main/kotlin/azure/BlobAccess.kt [hunk 5]
```diff
@@ -234,6 +242,75 @@ class BlobAccess() : Logging {
...
+        enum class SchemaType {
+            FHIR_FHIR,
+            FHIR_HL7,
+        }
+
+        /**
+         * Downloads the content of each blob in the [directory] with [blobContainerInfo] takes the input file, runs
+         * the transform on it, and compares it to the output validating that they are the same.
+         */
+        fun validateSchemas(
```
Comment (arnejduranovic) [inline prime-router/src/main/kotlin/azure/BlobAccess.kt:254]: This seems to be one of a few upcoming functions directly related to the "Manage Translation Schemas" feature. It would be good to move this function into another class (along with the future ones) since these functions do not implement "BlobAccess" but rather make use of it.

**Review by arnejduranovic — CHANGES_REQUESTED** [review]

File: prime-router/src/main/kotlin/azure/BlobAccess.kt [hunk 6]
```diff
@@ -234,6 +242,75 @@ class BlobAccess() : Logging {
...
+            val transforms = blobs.filter { it.currentBlobItem.name.endsWith(".yml") }
+            transforms.forEach { currentTransform ->
+                val currentTransformName = currentTransform.currentBlobItem.name
+
+                val transformBlobClient = sourceContainer.getBlobClient(currentTransformName)
+                val transform = transformBlobClient.downloadContent()
+
+                val regex = Regex("(.*/).*")
+                val path = regex.find(currentTransformName)!!.groups[1]!!.value
+                val inputBlobClient = sourceContainer.getBlobClient(path + "input.fhir")
```
[the remark this answers is not in the document]
Comment (JessicaWNava) [reply prime-router/src/main/kotlin/azure/BlobAccess.kt:271]: Which part?

File: prime-router/src/main/kotlin/azure/BlobAccess.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (JessicaWNava) [reply prime-router/src/main/kotlin/azure/BlobAccess.kt:280]: How are we able to detect what kind of schema we are dealing with. I can change it to use like Report.format or something but we will need to update it anyways if we add CSV because there will be a different converter to use

File: prime-router/src/main/kotlin/azure/BlobAccess.kt [hunk 7]
```diff
@@ -234,6 +242,75 @@ class BlobAccess() : Logging {
...
+                        transform.toStream(),
+                        FhirTransformSchema::class.java
+                    )
+                    val fhirTransform = if (configSchema is FhirTransformSchema) {
+                        FhirTransformer(configSchema).transform(inputBundle)
+                    } else {
+                        throw InternalServerErrorException("Transform schema $currentTransformName not a FhirTransform")
+                    }
+                    val output = sourceContainer.getBlobClient(path + "output.fhir").downloadContent()
+                    // Unfortunately need to do this to get it to be the same format.
```
[the remark this answers is not in the document]
Comment (JessicaWNava) [reply prime-router/src/main/kotlin/azure/BlobAccess.kt:286]: It will not pass if I compare the decoded bundles.

[the remark this answers is not in the document]
Comment (JessicaWNava) [reply prime-router/src/main/kotlin/azure/BlobAccess.kt:286]: Yah, we can compare the decoded files.

File: prime-router/src/main/kotlin/azure/BlobAccess.kt (hunk 4 above)
Comment (JessicaWNava) [reply prime-router/src/main/kotlin/azure/BlobAccess.kt:295]: I do not understand this comment, let's discuss.

File: prime-router/src/main/kotlin/fhirengine/translation/TranslationSchemaManager.kt [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+            BlobAccess.listBlobs(schemaType.directory, blobContainerInfo, false)
+        val inputs = blobs.filter { it.currentBlobItem.name.startsWith("input.") }
+
+        inputs.forEach { currentInput ->
+            val inputDirectoryPath = Regex("(.*/).*").find(currentInput.currentBlobItem.name)!!.groups[1]!!.value
+            val input = sourceContainer.getBlobClient(inputDirectoryPath + "input.fhir")
+            val inputBundle = FhirTranscoder.decode(input.toString())
+
+            val transformBlob = blobs.filter {
+                it.currentBlobItem.name.startsWith(inputDirectoryPath) &&
```
Comment (arnejduranovic) [inline prime-router/src/main/kotlin/fhirengine/translation/TranslationSchemaManager.kt:39]: I don' t think this logic will work. Looking at the design note, the structure shall look something like this:

```
/azure-blob-container
    /fhir_mapping
        valid-2023-03-29T23:53:00.488Z.txt
        previous-valid-2023-03-25T23:53:00.488Z.txt
    /fhir_transforms
        valid-2023-03-29T23:53:00.488Z.txt
        previous-valid-2023-03-25T23:53:00.488Z.txt
        /common
        /receivers
        /senders
    /hl7_mapping
        valid-2023-03-29T23:53:00.488Z.txt
        previous-valid-2023-03-25T23:53:00.488Z.txt
        /receivers
            /STLTS
                /CA
                    CA.yml
                    /resources
                        ca-aoe-note.yml
                        ...
                    input.fhir
                    output.hl7
            /flexion
                ...
```

IF we were to grab the input.fhir shown above, this logic will return multiple .yml files? Instead of always expecting the first result to be the one we want, may be better to match a regex that looks for exactly `inputDirectorypath/<someStringWithNoSlash>.yml`?

File: prime-router/src/main/kotlin/fhirengine/translation/TranslationSchemaManager.kt [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+                        FhirTransformer(configSchema).transform(inputBundle)
+                    } else {
+                        throw InternalServerErrorException("Transform schema $transformBlob not a FhirTransform")
+                    }
+                    val output = sourceContainer.getBlobClient(inputDirectoryPath + "output.fhir").downloadContent()
+                    // Unfortunately need to do this to get it to be the same format.
+                    val expectedOutput = FhirTranscoder.decode(output.toString())
+                    // .equals is required here == does not compare it properly, despite what IntelliJ says
+                    if (transformedInputBundle.equals(expectedOutput)) {
+                        BlobAccess.logger.error("Validation failed for transform $transformBlobName")
```
Comment (arnejduranovic) [inline prime-router/src/main/kotlin/fhirengine/translation/TranslationSchemaManager.kt:62]: Why are we using the BlobAccess logger?

File: prime-router/src/main/kotlin/fhirengine/translation/TranslationSchemaManager.kt [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+                    val output = sourceContainer.getBlobClient(inputDirectoryPath + "output.hl7").downloadContent()
+                    if (!hl7Transform.toString().trim().equals(output.toString().trim())) {
+                        BlobAccess.logger.error("Validation failed for transform $transformBlobName")
+                        return false
+                    }
+                }
+            }
+        }
+
+        return true
```
Comment (arnejduranovic) [inline prime-router/src/main/kotlin/fhirengine/translation/TranslationSchemaManager.kt:89]: I thought we discussed returning a data object that contains results of the validations? That way when running from the CLI, we can print out the validation results instead of making people dig through logs.

File: prime-router/src/main/kotlin/fhirengine/translation/TranslationSchemaManager.kt [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+            val input = sourceContainer.getBlobClient(inputDirectoryPath + "input.fhir")
+            val inputBundle = FhirTranscoder.decode(input.toString())
+
+            val transformBlob = blobs.filter {
+                it.currentBlobItem.name.startsWith(inputDirectoryPath) &&
+                it.currentBlobItem.name.endsWith(".yml")
+            }
+            val transformBlobName = transformBlob[0].currentBlobItem.name
+
+            when (schemaType) {
```
Comment (arnejduranovic) [inline prime-router/src/main/kotlin/fhirengine/translation/TranslationSchemaManager.kt:44]: No default case in this switch? In the case where the schemaType is something we don't support, we should figure out how to communicate that. I'm leaning towards updating the proposed data object to record the schemaType was not supported.

File: prime-router/src/test/kotlin/azure/resources/validationTests/FHIR_to_FHIR/input.fhir [hunk 12]
```diff
@@ -0,0 +1,13618 @@
+{
```
Comment (arnejduranovic) [inline prime-router/src/test/kotlin/azure/resources/validationTests/FHIR_to_FHIR/input.fhir:1]: Why are these under the azure folder when they are used in tests under the fhirengine folder? Seems like a better place would be `src/test/resources`?

File: prime-router/src/test/kotlin/fhirengine/engine/FhirConverterTests.kt [hunk 13]
```diff
@@ -294,6 +294,10 @@ class FhirConverterTests {
     fun `test getTransformerFromSchema`() {
         val engine = spyk(makeFhirEngine(metadata, settings, TaskAction.process) as FHIRConverter)
 
+        mockkClass(BlobAccess::class)
```
Comment (arnejduranovic) [inline prime-router/src/test/kotlin/fhirengine/engine/FhirConverterTests.kt:297]: Why are these changes needed now?

**Review by arnejduranovic — CHANGES_REQUESTED** [review]

File: prime-router/src/main/kotlin/fhirengine/translation/TranslationSchemaManager.kt (hunk 11 above)
Comment (JessicaWNava) [reply prime-router/src/main/kotlin/fhirengine/translation/TranslationSchemaManager.kt:44]: The when is exhaustive so there isn't a default case needed.

File: prime-router/src/test/kotlin/fhirengine/engine/FhirConverterTests.kt (hunk 13 above)
Comment (JessicaWNava) [reply prime-router/src/test/kotlin/fhirengine/engine/FhirConverterTests.kt:297]: Because we are passing the blob connection info around now and so the things that use those methods need that call mocked otherwise they fail.

**Comment by arnejduranovic** [comment]
This now seems to be a duplicate of #13346, so closing for now.
